### DISCUSSION TITLE
The Motion On Lathe is not Smooth ( delayed motion )

### DISCUSSION THREAD

**Discussion by Reprosense** [grblHAL/core#450] [General]
https://github.com/grblHAL/core/assets/135573322/59b7f3b5-3b20-4230-aff3-2557cc998b9e

hello guys , i'd like to ask , i use STM32F446 nucleo , and Protoneer v3 for the config , i try using lathe on this GRBLHal , and from my assumption its happen because of the CSS ( G96 ) , its speed ( feedrate ) moved based on the speed of encoder right ? ,, its show that sometime the value of encoder got 0 ( idk yet , is this because of the calculation of css or there is any delay ) when its change the line between the gcode , or did u have any clue ? 

here is my settings
[VER:1.1f.20240127:]
[OPT:VNZH0SL,35,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,HOME,NOPROBE,ES,LATHE,SS,SED]
[FIRMWARE:grblHAL]
[SIGNALS:HSEFM]
[NVS STORAGE:*FLASH]
[FREE MEMORY:111K]
[DRIVER:STM32F446]
[DRIVER VERSION:240125]
[BOARD:Protoneer v3]
[AUX IO:0,2,0,0]
ok
>>> $$
$0 = 4.0    (Step pulse time, microseconds)
$1 = 25    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 0    (Step direction invert, mask)
$4 = 7    (Invert step enable pin, boolean)
$5 = 7    (Invert limit pins, boolean)
$9 = 3   
$10 = 511    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.001    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$14 = 6   
$15 = 0   
$16 = 0   
$17 = 0   
$18 = 0   
$20 = 1    (Soft limits enable, boolean)
$21 = 0    (Hard limits enable, boolean)
$22 = 11    (Homing cycle enable, boolean)
$23 = 0    (Homing direction invert, mask)
$24 = 100.0    (Homing locate feed rate, mm/min)
$25 = 1000.0    (Homing search seek rate, mm/min)
$26 = 250    (Homing switch debounce delay, milliseconds)
$27 = 4.000    (Homing switch pull-off distance, millimeters)
$28 = 0.100   
$29 = 0.0   
$30 = 4000.600    (Maximum spindle speed, RPM)
$31 = 5.000    (Minimum spindle speed, RPM)
$32 = 2    (Laser-mode enable, boolean)
$33 = 5000.0   
$34 = 0.0   
$35 = 0.0   
$36 = 100.0   
$37 = 0   
$38 = 1024   
$39 = 1   
$40 = 0   
$43 = 1   
$44 = 1   
$45 = 4   
$46 = 0   
$62 = 0   
$63 = 3   
$64 = 0   
$66 = ,,   
$67 = ,,   
$68 = ,,   
$69 = ,,   
$90 = 0.000   
$91 = 0.000   
$92 = 0.000   
$95 = 0.000   
$100 = 500.00000    (X-axis travel resolution, step/mm)
$101 = 500.00000    (Y-axis travel resolution, step/mm)
$102 = 500.00000    (Z-axis travel resolution, step/mm)
$110 = 8000.000    (X-axis maximum rate, mm/min)
$111 = 8000.000    (Y-axis maximum rate, mm/min)
$112 = 8000.000    (Z-axis maximum rate, mm/min)
$120 = 500.000    (X-axis acceleration, mm/sec^2)
$121 = 500.000    (Y-axis acceleration, mm/sec^2)
$122 = 500.000    (Z-axis acceleration, mm/sec^2)
$130 = 300.000    (X-axis maximum travel, millimeters)
$131 = 200.000    (Y-axis maximum travel, millimeters)
$132 = 200.000    (Z-axis maximum travel, millimeters)
$340 = 50.0   
$341 = 0   
$342 = 30.0   
$343 = 25.0   
$344 = 200.0   
$345 = 200.0   
$346 = 1   
$370 = 0   
$372 = 0   
$384 = 0   
$398 = 35   
$481 = 0   
$484 = 1   
$486 = 0   
ok
>>> $pins
[PIN:PA0,Emergency stop]
[PIN:PA1,Feed hold]
[PIN:PA4,Cycle start]
[PIN:PC7,X limit min]
[PIN:PB6,Y limit min]
[PIN:PA6,Z limit min]
[PIN:PC9,Spindle index]
[PIN:PA11,Aux input 0,Motor warning]
[PIN:PA12,Aux input 1,Motor fault]
[PIN:PA10,X step]
[PIN:PB3,Y step]
[PIN:PB5,Z step]
[PIN:PB4,X dir]
[PIN:PB10,Y dir]
[PIN:PA8,Z dir]
[PIN:PA9,Steppers enable]
[PIN:PB7,Spindle on]
[PIN:PA5,Spindle direction]
[PIN:PB0,Flood]
[PIN:PB15,Aux out 0,P0]
[PIN:PC12,Aux out 1,P1]
[PIN:PA3,RX,Primary UART]
[PIN:PA2,TX,Primary UART]
[PIN:PA7,Spindle PWM]
[PIN:PD2,Spindle pulse]

here is my gcode [too]([the Gcode.txt](https://github.com/grblHAL/core/files/14391458/the.Gcode.txt))

thanks in advace

**Comment by terjeio** [comment]
> ... from my assumption its happen because of the CSS ( G96 ) , its speed ( feedrate ) moved based on the speed of encoder right ?

No, when[ CSS is enabled](http://linuxcnc.org/docs/2.6/html/gcode/gcode.html#sec:G96-G97-Spindle) the spindle RPM is adjusted based on how far the tool is from the center (0) of the workpiece.
If you do not use CSS is the movement smooth?
And I wonder if the S200 argument for G96 is a bit on the high side for such a small diameter part since I cannot hear the RPM changing with the X position.

BTW I need a bit of time to look into this as it is not easy to simulate witch a bench setup - I believe Ihave to check it out in the workshop.

**Reply by terjeio** [reply]
You have G95 enabled as well and that indeed requires input from the encoder. And it requires the [PID parameters to be tuned](https://github.com/grblHAL/core/wiki/Spindle-sync) which I can see from the settings has not be done.

**Reply by Reprosense** [reply]
okay terje , ill try to tuned and comeback later

**Reply by terjeio** [reply]
I have tested with your gcode and found a few issues. First is that arcs (G2/G3) in G96 mode runs very slow due to a missing parameter in the command sent to motion control. The jerky motion seems to be due to spindle RPM changes beeing accompanied by M3 - which for Grbl/grblHAL causes the planner to be emptied plus possible the spindle at speed routine waiting for the spindle to reach its programmed rate.
The first is easy to fix, the second needs a bit of tought and testing.

**Reply by terjeio** [reply]
I have comitted a [tentative fix](https://github.com/grblHAL/core/commit/06f442643a115ab3acd6722b4afc1fcee8d9d7be), your file seems to work ok with my lathe.
G95 uses spindle synced motion which requires tuning of the PID parameters, however I am wondering if I should add a setting to disable PID feedback and just use the RPM input. I believe that would work since there is no need (?) to start passes at the exact same spot as threading requires. G95 mode cannot be used for threading anyway so no loss of functionality.

BTW the use of G96 in your file is a bit odd, most set the max feed rate as the same required for the start postion so no, or very small, RPM changes will occur. May I ask which CAM are you using?

**Reply by Reprosense** [reply]
for the CAM , im using Fusion and generate it with LinuxCNC turning
